pkg/nsx/services/ippool/ippool.go

```diff
@@ -0,0 +1,251 @@
+	if !(IPPoolUpdated || IPPoolSubnetsUpdated) {
+		return nil
+	}
+	infraIPPool, err := service.WrapHierarchyIPPool(nsxIPPool, nsxIPSubnets)
+	if err != nil {
+		return err
+	}
+	// Get IPPool Type from nsxIPPool
+	IPPoolType := common.IPPoolTypePublic
+	for _, tag := range nsxIPPool.Tags {
+		if *tag.Scope == common.TagScopeIPPoolCRType {
+			IPPoolType = *tag.Tag
+			break
+		}
+	}
+
+	if IPPoolType == common.IPPoolTypePrivate {
+		ns := service.GetIPPoolNamespace(nsxIPPool)
+		VPCInfo := commonctl.ServiceMediator.GetVPCInfo(ns)
+		if len(VPCInfo) == 0 {
+			err = util.NoEffectiveOption{Desc: "no effective org and project for ippool"}
+		} else {
+			err = service.NSXClient.ProjectInfraClient.Patch(VPCInfo[0].OrgID, VPCInfo[0].ProjectID, *infraIPPool,
+				&EnforceRevisionCheckParam)
+		}
+	} else if IPPoolType == common.IPPoolTypePublic {
+		err = service.NSXClient.InfraClient.Patch(*infraIPPool, &EnforceRevisionCheckParam)
+	} else {
+		err = util.NoEffectiveOption{Desc: "no effective ippool type"}
+	}
+	if err != nil {
+		return err
+	}
+	if IPPoolUpdated {
+		err = service.ipPoolStore.Operate(nsxIPPool)
+		if err != nil {
+			return err
+		}
+	}
+	if IPPoolSubnetsUpdated {
+		err = service.ipPoolBlockSubnetStore.Operate(nsxIPSubnets)
+		if err != nil {
+			return err
+		}
+	}
+	log.V(1).Info("successfully created or updated ippool and ip subnets", "nsxIPPool", nsxIPPool)
+	return nil
+}
+
+func (service *IPPoolService) AcquireRealizedSubnetIP(obj *v1alpha2.IPPool) ([]v1alpha2.SubnetResult, bool, error) {
+	realizedSubnets := []v1alpha2.SubnetResult{}
+	subnetCidrUpdated := false
+	for _, subnetRequest := range obj.Spec.Subnets {
+		// check if the subnet is already realized
+		realized := false
+		realizedSubnet := v1alpha2.SubnetResult{Name: subnetRequest.Name}
+		for _, statusSubnet := range obj.Status.Subnets {
+			if statusSubnet.Name == subnetRequest.Name && statusSubnet.CIDR != "" {
+				realizedSubnet.CIDR = statusSubnet.CIDR
+				realized = true
+				break
+			}
+		}
+		if !realized {
+			cidr, err := service.acquireCidr(obj, &subnetRequest, common.RealizeMaxRetries)
+			if err != nil {
+				return nil, subnetCidrUpdated, err
+			}
+			if cidr != "" {
+				subnetCidrUpdated = true
+			}
+			realizedSubnet.CIDR = cidr
+		}
+		realizedSubnets = append(realizedSubnets, realizedSubnet)
+	}
+	return realizedSubnets, subnetCidrUpdated, nil
+}
+
+func (service *IPPoolService) DeleteIPPool(obj interface{}) error {
+	var err error
+	var nsxIPPool *model.IpAddressPool
+	nsxIPSubnets := make([]*model.IpAddressPoolBlockSubnet, 0)
+	switch o := obj.(type) {
+	case *v1alpha2.IPPool:
+		nsxIPPool, nsxIPSubnets = service.BuildIPPool(o)
+		if err != nil {
+			log.Error(err, "failed to build ip pool", "IPPool", o)
+			return err
+		}
+	case types.UID:
+		nsxIPPool, nsxIPSubnets, err = service.indexedIPPoolAndIPPoolSubnets(o)
+		if err != nil {
+			log.Error(err, "failed to get ip pool and ip pool subnets by UID", "UID", o)
```

Comment:
if ip pool found, but ip pool subnet not found,  or just subnet found, no ip pool found, do we also need to delete resources?

Reply:
we can take them as independent resources currently.